pkg/git/copy_test.go

```diff
@@ -29,18 +30,77 @@ func TestCopyService(t *testing.T) {
 	}
 }
 
+func TestPathValidForPromotion(t *testing.T) {
+
+	serviceBeingPromoted := "service-name"
+	servicesNotBeingPromoted := []string{"svc", "base", "serviceName", "services"}
+	promoteTheseWhenServiceNameIsRight := []string{
+		"services/service-name/base/config/kustomization.yaml",
+		"services/service-name/base/config/deployment.yaml",
+		"services/service-name/base/config/dir/below/it/may/contain/important.yaml",
+	}
+	for _, filePath := range promoteTheseWhenServiceNameIsRight {
+		if !pathValidForPromotion(serviceBeingPromoted, filePath) {
+			t.Fatalf("Valid path for promotion for %s incorrectly rejected: %s", serviceBeingPromoted, filePath)
+		}
+		for _, wrongService := range servicesNotBeingPromoted {
+			if pathValidForPromotion(wrongService, filePath) {
+				t.Fatalf("Path for service %s incorrectly accepted for promotion: %s", wrongService, filePath)
+			}
+		}
+	}
+
+	// These paths should never be promoted
+	badServiceNames := []string{"svc", "badService"}
+	neverPromoteThese := []string{
+		"services/svc/overlays/kustomization.yaml",
+		"services/svc/kustomization.yaml",
+		"svc/base/config/deployment.yaml",
+		"services/badService/any/other/path/resource.yaml",
+	}
+	for _, badPath := range neverPromoteThese {
+		for _, badServiceName := range badServiceNames {
+			if pathValidForPromotion(badServiceName, badPath) {
+				t.Fatalf("Invalid path %s for promotion of service %s incorrectly accepted", badPath, badServiceName)
+			}
+		}
+	}
+}
+
```

Comment:
There's probably a missing test here which would highlight a bug.

```go
func TestCopyServiceWithFailureCopying(t *testing.T) {
	testError := errors.New("this is a test error")
	s := &mockSource{localPath: "/tmp/testing"}
	files := []string{"services/service-a/base/config/my-file.yaml", "services/service-a/base/config/this-file.yaml"}
	for _, f := range files {
		s.addFile(f)
	}
	d := &mockDestination{}
	d.copyError = testError

	copied, err := CopyService("service-a", s, d)
	if err != testError {
		t.Fatalf("unexpected error: got %v, wanted %v", err, testError)
	}
	d.assertFilesWritten(t, []string{})
	if !reflect.DeepEqual(copied, []string{}) {
		t.Fatalf("failed to copy the files, got %#v, want %#v", copied, []string{})
	}
}
```

I'd rewrite `mockDestination`'s CopyFile (and add mockDestination.copyError) as:

```go
func (d *mockDestination) CopyFile(src, dst string) error {
	if d.written == nil {
		d.written = []string{}
	}
	if d.copyError != nil {
		return d.copyError
	}
	d.written = append(d.written, dst)
	return nil
}
```

Reply:
What a great catch, thank you! Fix is pushed.